kernel/src/snake.rs

```diff
@@ -0,0 +1,375 @@
+        Grid {
+            cells: vec![Cell::Empty; width * height],
+            width,
+            height,
+        }
+    }
+
+    fn clear(&mut self) {
+        for cell in self.cells.iter_mut() {
+            *cell = Cell::Empty;
+        }
+    }
+
+    fn set(&mut self, point: Point, cell: Cell) {
+        self.cells[Grid::index(point)] = cell;
+        STDOUT.write().set_char(cell.character(), point)
+            .expect("failed to draw cell to screen");
+    }
+
+    fn get(&self, point: Point) -> &Cell {
+        &self.cells[Grid::index(point)]
+    }
+
+    #[inline]
+    fn index(point: Point) -> usize {
+        if point.x >= 80 || point.y >= 25 {
+            panic!("point out of bounds {} {}", point.x, point.y);
+        }
+        point.x + point.y * 80
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum Cell {
+    Empty,
+    Head,
+    Body,
+    Apple,
+}
+
+impl Cell {
+    fn character(&self) -> TerminalCharacter {
+        match self {
+            Cell::Empty => TerminalCharacter::new(' ', color!(Black on Black)),
+            Cell::Head => TerminalCharacter::new(HEAD_CHAR, color!(White on Black)),
+            Cell::Body => TerminalCharacter::new(' ', color!(Green on Green)),
+            Cell::Apple => TerminalCharacter::new(' ', color!(Red on Red))
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+enum Direction {
+    Up,
+    Down,
+    Left,
+    Right,
+}
+
+struct Snake {
+    head: Point,
+    direction: Direction,
+    blocks: Vec<Point>,
+    len: u16,
+}
+
+impl Snake {
+    fn new() -> Snake {
+        Snake {
+            head: STDOUT.read().resolution().center(),
+            direction: Direction::Right,
+            blocks: Vec::with_capacity(128),
+            len: BASE_LENGTH,
+        }
+    }
+
+    fn update(&mut self, grid: &mut Grid) -> MoveResult {
+        // Replace the previous head position with a snake body block
+        if !self.blocks.is_empty() {
+            grid.set(self.head, Cell::Body);
+        }
+
+        let move_result = self.try_move(grid);
+        match move_result {
+            MoveResult::Moved(point) => self.head = point,
+            MoveResult::Win => return MoveResult::Win,
+            MoveResult::Lose => return MoveResult::Lose,
+        }
+
+        // Draw the head
+        grid.set(self.head, Cell::Head);
+
+        // Update the train
+        if self.blocks.len() < self.len as usize {
+            self.blocks.insert(0, self.head);
+        } else {
+            // Clear the tail point
+            if let Some(tail) = self.blocks.last() {
+                grid.set(tail.clone(), Cell::Empty);
+            }
+
+            self.blocks.rotate_right(1);
+            self.blocks[0] = self.head;
+        }
+
+        move_result
+    }
+
+    fn try_move(&mut self, grid: &mut Grid) -> MoveResult {
+        let moved_head = self.direction.offset(self.head, 1);
+        if !STDOUT.read().in_bounds(moved_head) {
```

Comment:
Maybe should check based on grid width & height? May not be the same

Reply:
They are for now though

Reply:
Sure, but in theory it is possible. It is better we fix the code now than change something and break it later, imo :P